ydb/query/base.py

```diff
@@ -0,0 +1,332 @@
+    session's lifecycle manually - use a QuerySessionPool is always a better choise.
+    """
+
+    @abc.abstractmethod
+    def __init__(self, driver: SupportedDriverType, settings: QueryClientSettings = None):
+        pass
+
+    @abc.abstractmethod
+    def create(self) -> "IQuerySession":
+        """
+        Creates a Session of Query Service on server side and attaches it.
+
+        :return: Session object.
+        """
+        pass
+
+    @abc.abstractmethod
+    def delete(self) -> None:
+        """
+        Deletes a Session of Query Service on server side and releases resources.
+
+        :return: None
+        """
+        pass
+
+    @abc.abstractmethod
+    def transaction(self, tx_mode: BaseQueryTxMode) -> "IQueryTxContext":
+        """
+        Creates a transaction context manager with specified transaction mode.
+
+        :param tx_mode: Transaction mode, which is a one from the following choises:
+         1) QuerySerializableReadWrite() which is default mode;
+         2) QueryOnlineReadOnly(allow_inconsistent_reads=False);
+         3) QuerySnapshotReadOnly();
+         4) QueryStaleReadOnly().
+
+        :return: transaction context manager.
+        """
+        pass
+
+
+class IQueryTxContext(abc.ABC):
+    """
+    An object that provides a simple transaction context manager that allows statements execution
+    in a transaction. You don't have to open transaction explicitly, because context manager encapsulates
+    transaction control logic, and opens new transaction if:
+     1) By explicit .begin();
+     2) On execution of a first statement, which is strictly recommended method, because that avoids
+     useless round trip
+
+    This context manager is not thread-safe, so you should not manipulate on it concurrently.
+    """
+
+    @abc.abstractmethod
+    def __init__(
+        self,
+        driver: SupportedDriverType,
+        session_state: IQuerySessionState,
+        session: IQuerySession,
+        tx_mode: BaseQueryTxMode = None,
+    ):
+        pass
+
+    @abc.abstractmethod
+    def __enter__(self):
+        """
+        Enters a context manager and returns a transaction
+
+        :return: A transaction instance
+        """
+        pass
+
+    @abc.abstractmethod
+    def __exit__(self, *args, **kwargs):
+        """
+        Closes a transaction context manager and rollbacks transaction if
+        it is not rolled back explicitly
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def session_id(self):
+        """
+        A transaction's session id
+
+        :return: A transaction's session id
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def tx_id(self):
+        """
+        Returns a id of open transaction or None otherwise
+
+        :return: A id of open transaction or None otherwise
+        """
+        pass
+
+    @abc.abstractmethod
+    def begin(settings: QueryClientSettings = None):
```

Comment:
here and below: 
1. first parameter should be self
2. if default value is None - type hint must respect it and be Optional